### ISSUE TITLE
Long conversion time

### ISSUE THREAD

**Issue by Mike-DE-RE** [RWAP/PrinterToPDF#9]
The conversion time (especially from png to pdf) remains very long:

time ./PrinterConvert 4 3 SIEMENS.C16 1 sdloff ./

delays around ack: t1=0    t2=0    t3=0    t4=0    t5=0
write   = png/page5.png
command = convert  png/page5.png  ./pdf/page5.pdf

I am at page 6
write   = png/page6.png
command = convert  png/page6.png  ./pdf/page6.pdf

real    20m39.412s
user    0m59.700s
sys     0m15.925s

**Comment by Mike-DE-RE** [comment]
The conversion time has improved. I've added more RAM to my PC. ;-)

With 2GB is it better but still far away from acceptable.

time ./PrinterConvert 4 3 SIEMENS.C16 1 sdloff ./

delays around ack: t1=0    t2=0    t3=0    t4=0    t5=0
write   = png/page1.png
command = convert  png/page1.png  ./pdf/page1.pdf
write   = png/page2.png
command = convert  png/page2.png  ./pdf/page2.pdf


I am at page 3
write   = png/page3.png
command = convert  png/page3.png  ./pdf/page3.pdf

real    2m15.413s
user    1m21.221s
sys     0m6.952s

This is the time for RETRO.zip sample.

**Comment by RWAP** [comment]
I have added a check in the write_png routine to ignore any blank pages.

One speed-up would be to create a table to lookup the RGB values for each value fetched from lookupColour, so that it is only called once per colourValue stored in the printermemory.  This should improve the speed of generating the PNG image.  Open to suggestions as to how to do this.

**Comment by Mike-DE-RE** [comment]
Can't that be calculated or at least approximated? You would not need a table (memory). To be honest, I do not understand what you are doing in _lookupColour_.

**Comment by RWAP** [comment]
I have now implemented a new method - it is possibly not the best solution, but I have installed the Libharu PDF library http://libharu.org/.  Whilst this is only slightly quicker than ImageMagick for converting a PNG file to PDF, it does have the benefit that you can insert a raw image file into the PDF direct from memory.

This does unfortunately mean that the application needs to set aside a large amount of memory to hold the image file (3 x pageWidth x pageHeight bytes), unless we change the RGB to use 8 bits instead of 24 bits.

Speed improvement is noticeable particularly in creating the actual image for insertion into the PDF (approx 1/3 of the time).  It still takes 20 seconds to get that image into the PDF though....

**Comment by Mike-DE-RE** [comment]
This is my test using the newly implemented method:

time ./PrinterConvert 4 3 SIEMENS.C16 1 sdloff ./

delays around ack: t1=0    t2=0    t3=0    t4=0    t5=0
Start PNG : 1495197609 1495197610 1
write   = png/page1.png
End PNG : 1495197609 1495197612 3


I am at page 2
Start PNG : 1495197609 1495197625 16
write   = png/page2.png
End PNG : 1495197609 1495197627 18

real    0m25.161s
user    0m18.781s
sys     0m1.724s

**Comment by Mike-DE-RE** [comment]
Same conversion with old method (ImageMagic - convert):

time ./PrinterConvert 4 3 SIEMENS.C16 1 sdloff ./

delays around ack: t1=0    t2=0    t3=0    t4=0    t5=0
write   = png/page1.png
command = convert  png/page1.png  ./pdf/page1.pdf


I am at page 2
write   = png/page2.png
command = convert  png/page2.png  ./pdf/page2.pdf

real    0m50.326s
user    0m43.127s
sys     0m3.396s

**Comment by Mike-DE-RE** [comment]
I've simply commented out

lines 762-763
// sprintf(filenameX, "convert  %spage%d.png  %spage%d.pdf ", pathpng,
//    page, pathpdf, page);

and

line 3820
// sprintf(filenameX, "convert  %spage%d.png  %spage%d.pdf ", pathpng, page,pathpdf, page);

time ./PrinterConvert 4 3 SIEMENS.C16 1 sdloff ./

delays around ack: t1=0    t2=0    t3=0    t4=0    t5=0
write   = png/page1.png
command = png/page1.png
sh: png/page1.png: Keine Berechtigung


I am at page 2
write   = png/page2.png
command = png/page2.png
sh: png/page2.png: Keine Berechtigung

real    0m29.715s
user    0m25.658s
sys     0m0.632s

I have no idea about the shell Errors.

**Conclusion:**

New method is even faster than pure png creation.

If converted both png files to TIF using Gimp (one file is completely empty). Then I copied both single TIF files in one file using tiffcp (multi page TIF) and converted this to PDF:

time tiff2pdf collect.tif -o page.pdf -z

real    0m18.862s
user    0m15.373s
sys     0m0.948s

By simple calculation this would mean a total time of

real    0m48.577s
user    0m41.031s
sys     0m1.580s

for the process generating a PNG and converting a TIF to PDF.

Comparing it to above test

> real 0m25.161s
> user 0m18.781s
> sys 0m1.724s

The result is quite clear.

Shorter time would even be better, but cannot be reached the way described above.
